### DISCUSSION TITLE
Contrast-ratio problems

### DISCUSSION THREAD

**Discussion by 7iomka** [mantinedev/mantine#1727] [Q&A]
Hi.
I changed `amber` color palette to this

```
['#fffbf0', '#fff8e5', '#fff1cc', '#ffe9b3', '#ffe194', '#ffda7c', '#ffc42e', '#e6a400', '#996e00', '#4d3700']
```

and in provider I settled
```
 primaryColor: 'amber',
 primaryShade: 5,
```

That is  now looks `filled` variant of button

<img width="176" alt="image" src="https://user-images.githubusercontent.com/18473137/177041961-7f1ad1d4-b3d7-4721-9b0d-25319d5dafa0.png">

[Contrast ratio](https://webaim.org/resources/contrastchecker/) for `color: #fff` when backgroud is `#FFDA7C;` is `1.34:1` and of course it fails on all WCAG tests.

That is happening now if I use all filled components
<img width="411" alt="image" src="https://user-images.githubusercontent.com/18473137/177042094-8c68f068-e4c9-4a0b-a251-1b6eefcbeb4d.png">

Our designer know this problem, and, when it is needed he change text color from white to black (or similar opposite).
<img width="340" alt="image" src="https://user-images.githubusercontent.com/18473137/177042156-3e98f7da-9cfb-43c6-9a4d-ca5f7a161c1a.png">
<img width="264" alt="image" src="https://user-images.githubusercontent.com/18473137/177042183-74982568-17af-4dfc-9a19-cdfd6681fd82.png">
<img width="150" alt="image" src="https://user-images.githubusercontent.com/18473137/177042199-c9d75c6e-0a77-4564-b9f6-572da3e9a5f4.png">

My question is - how to configure globally text-colors, inner background, etc, when second color now is hardcoded to `white`?

There we have 2 options for a feature request:
- autoconfigure aspect ratio
- autoconfigure aspect ratio with ability to partial override

When I worked with css-variables, solution for this was really simple - inside config we declare inverse color for each color manually.
<img width="348" alt="image" src="https://user-images.githubusercontent.com/18473137/177042351-e02eeb90-878d-480e-9539-cd76090e3ac0.png">
<img width="679" alt="image" src="https://user-images.githubusercontent.com/18473137/177042429-72af0919-a0ee-41ba-a189-48ec159098f5.png">

**Comment by rtivital** [comment]
You can use styles params on MantineProvider to change color of buttons with color – https://mantine.dev/theming/mantine-provider/#styles-on-mantineprovider

**Comment by 7iomka** [comment]
@rtivital I now this, but it's hard to do and hard to maintain.
After all, we need for each color of the palette (which should have a second/opposite color other than `white` for reasons of contrast) to create a configuration for each component where this color is used in combination.

**Reply by rtivital** [reply]
I cannot think of any other solution that is available right now, can you propose how this should be handled on library level?

**Comment by 7iomka** [comment]
@rtivital 

On library level default behavior for all components, that now has hardcoded `white` color I would like to see that setup

It is just an example, how components needed to be configured by default on library level internally.
If user won't like this behavior, it can be disabled for example by option: `theme.enableReadableOppositeColors`:

```tsx
import { mostReadable } from '@ctrl/tinycolor';  // can be used internally and maybe exposed to `theme.fn.mostReadableColor`

const getReadableColor = ({
  color,
  colorVariants = ['#fff', '#000'],
  isEnabled = true,
}: {
  color: string;
  colorVariants: string[];
  isEnabled: boolean;
}): string | undefined => (isEnabled ? mostReadable(color, colorVariants)?.toHexString() : color);


const themeConfig: MantineThemeOverride = {
  colorScheme: 'light',
  breakpoints,
  colors: mantineColors,
  primaryColor: 'amber',
  primaryShade: 5,
  components: {
    Button: {
      styles: (theme: MantineTheme, params: ButtonStylesParams) => ({
        root: {
          color:
            params.variant === 'filled'
              ? getReadableColor({
                  color: theme.colors[params.color || theme.primaryColor][theme.primaryShade],
                  isEnabled: theme.enableReadableOppositeColors,
                })
              : undefined,
        },
      }),
    },
    Checkbox: {
      // Subscribe to theme and component params
      styles: (theme: MantineTheme, params: ButtonStylesParams) => ({
        icon: {
          color: getReadableColor({
            color: theme.colors[params.color || theme.primaryColor][theme.primaryShade],
            isEnabled: theme.enableReadableOppositeColors,
          }),
        },
      }),
    },
    Radio: {
      // Subscribe to theme and component params
      styles: (theme: MantineTheme, params: ButtonStylesParams) => ({
        icon: {
          color: getReadableColor({
            color: theme.colors[params.color || theme.primaryColor][theme.primaryShade],
            isEnabled: theme.enableReadableOppositeColors,
          }),
        },
      }),
    },
  },
};
```
